### PR TITLE
Java client connection state part 1

### PR DESCRIPTION
**Pull request by BrennanConroy**
I'm limiting this change to just adding the `CONNECTING` state and modifying a couple things so it works as expected.

Part 2 will fix things like calling stop while in the middle of start, but I want to clean up the code a bit before doing that, so I'm planning on making a "clean up" PR before doing Part 2.

Easier to review if you turn off whitespace changes.

Progress towards https://github.com/dotnet/aspnetcore/issues/23291

### CODE REVIEW

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 1]
```diff
@@ -324,79 +326,93 @@ public void setBaseUrl(String url) {
...
-        });
 
-        stopError = null;
-        Single<NegotiateResponse> negotiate = null;
-        if (!skipNegotiate) {
-            negotiate = tokenCompletable.andThen(Single.defer(() -> startNegotiate(baseUrl, 0)));
-        } else {
-            negotiate = tokenCompletable.andThen(Single.defer(() -> Single.just(new NegotiateResponse(baseUrl))));
-        }
+            hubConnectionState = HubConnectionState.CONNECTING;
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:337]: We should do verbose logging for each state transition.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 2]
```diff
@@ -324,79 +326,93 @@ public void setBaseUrl(String url) {
...
-                        return Completable.complete();
+                            return Completable.complete();
+                        }));
                     }));
                 }));
-            }));
-        // subscribe makes this a "hot" completable so this runs immediately
-        }).subscribeWith(start);
+            // subscribe makes this a "hot" completable so this runs immediately
+            }).subscribe(() -> {
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:405]: The CONNECTING->CONNECTED state transition is hidden in this diff, but we should assert the previous state where we can.

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/HubConnectionTest.java [hunk 3]
```diff
@@ -1576,11 +1577,12 @@ public void receiveHandshakeResponseAndMessage() {
             value.getAndUpdate((val) -> val + 1);
         });
 
+        SingleSubject<String> handshakeMessageTask = mockTransport.getNextSentMessage();
         // On start we're going to receive the handshake response and also an invocation in the same payload.
         hubConnection.start();
-        mockTransport.getStartTask().timeout(1, TimeUnit.SECONDS).blockingAwait();
+        String sentMessage = handshakeMessageTask.timeout(1, TimeUnit.SECONDS).blockingGet();
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/HubConnectionTest.java:1583]: Do these short 1 second timeouts ever cause flakiness? If not, I'm impressed.

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/HubConnectionTest.java:1583]: There were one or two flaky tests in java but I don't think they were ever related to these timeouts.

And actually, yesterday I was looking into them and there is a pretty good reason these never trigger. We can talk about it later today.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 4]
```diff
@@ -324,79 +326,93 @@ public void setBaseUrl(String url) {
...
-        }
+        CompletableSubject start = CompletableSubject.create();
 
-        accessTokenProvider.subscribe(token -> {
-            if (token != null && !token.isEmpty()) {
-                this.localHeaders.put("Authorization", "Bearer " + token);
+        hubConnectionStateLock.lock();
+        try {
+            if (hubConnectionState != HubConnectionState.DISCONNECTED) {
+                return Completable.error(new RuntimeException("The HubConnection cannot be started if it is not in the 'Disconnected' state."));
```
Comment (BrennanConroy) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:334]: We should wait for the "in-progress" start here is there is one.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java (hunk 2 above)
Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:405]: I have a PR that extracts this into a method, I'll add the logging though

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 5]
```diff
@@ -426,7 +442,7 @@ public void run() {
     }
 
     private Single<NegotiateResponse> startNegotiate(String url, int negotiateAttempts) {
-        if (hubConnectionState != HubConnectionState.DISCONNECTED) {
+        if (hubConnectionState != HubConnectionState.CONNECTING) {
             return Single.just(null);
```
Comment (BrennanConroy) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:446]: throw here

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java [hunk 6]
```diff
@@ -341,83 +342,99 @@ public void setBaseUrl(String url) {
...
-        }
+        CompletableSubject localStart = CompletableSubject.create();
 
-        accessTokenProvider.subscribe(token -> {
-            if (token != null && !token.isEmpty()) {
-                this.localHeaders.put("Authorization", "Bearer " + token);
+        hubConnectionStateLock.lock();
+        try {
+            if (hubConnectionState != HubConnectionState.DISCONNECTED) {
+                logger.debug("Another start is in progress. Waiting for start to complete.");
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnection.java:350]: The hub could be already-started here too, right? Maybe log the current state.

**Review by halter73 — APPROVED** [review]
I don't love approving this without the changes to assert previous states and log state transitions, but I trust those are coming.

**Pull request merged**